src/components/views/settings/devices/DeviceDetailHeading.tsx

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { FormEvent, useEffect, useState } from 'react';
+
+import { _t } from '../../../../languageHandler';
+import AccessibleButton from '../../elements/AccessibleButton';
+import Field from '../../elements/Field';
+import Spinner from '../../elements/Spinner';
+import { Caption } from '../../typography/Caption';
+import Heading from '../../typography/Heading';
+import { DeviceWithVerification } from './types';
+
+interface Props {
+    device: DeviceWithVerification;
+    saveDeviceName: (deviceName: string) => Promise<void>;
+}
+
+const DeviceNameEditor: React.FC<Props & { stopEditing: () => void }> = ({
+    device, saveDeviceName, stopEditing,
+}) => {
+    const [deviceName, setDeviceName] = useState(device.display_name || '');
+    const [isLoading, setIsLoading] = useState(false);
+    const [error, setError] = useState<string | null>(null);
+
+    useEffect(() => {
+        setDeviceName(device.display_name || '');
+    }, [device.display_name]);
+
+    const onInputChange = (event: React.ChangeEvent<HTMLInputElement>): void =>
+        setDeviceName(event.target.value);
+
+    const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
+        setIsLoading(true);
+        setError(null);
+        event.preventDefault();
+        try {
+            await saveDeviceName(deviceName);
+            stopEditing();
+        } catch (error) {
+            setError(_t('Failed to set display name'));
+            setIsLoading(false);
+        }
+    };
+
+    const headingId = `device-rename-${device.device_id}`;
```

Comment:
We [should ensure](https://html.spec.whatwg.org/multipage/dom.html#the-id-attribute) that IDs don't contain spaces.